### DISCUSSION TITLE
Specify exact color palette for indexed pixel formats

### DISCUSSION THREAD

**Discussion by Solstice245** [ImageMagick/ImageMagick#6054] [General]
### Is your feature request related to a problem? Please describe.

I'm currently having trouble with getting BMPs out of a GIF file while preserving the palette data. I've googled around and it seems there is simply nothing out there available which is tailored to this particular problem. Technically, I was able to use GIMP to "solve" the problem, but GIMP is so slow that the volume of images which I need to run these conversions on would take quite literally hours as opposed to what would probably take this tool a handful of minutes.

### Describe the solution you'd like

It's very simple, I'd like the option when doing conversions such as `magick convert sample.gif -coalesce -colors 256 sample.bmp` to specify the palette, either taken directly from the source file or an external file, preferably `.pal` or other such file type. `-palette-source` and `-palette` arguments would do nicely. (Or whatever terms would seem more fitting)

**Comment by fmw42** [comment]
How about using -remap.  Get the color from the color map of the GIF and create a colormap image of them, one pixel per color appended together either horizontally or vertically. Then use -remap with the colormap image on your bmp or in your original command line after the -colors 256.  

Note, I am not sure why you need -colors 256 as GIF is limited to 256 colors. Also -colors 256 does not guarantee you will get 256 colors if there are fewer colors in the image.

**Reply by emk2203** [reply]
Could you give an example command for this? I am unclear about the parts in the middle.

**Comment by snibgo** [comment]
IM doesn't contain CLI options for general manipulation of palettes. However, if we have a simple command...
```
magick in.gif out.bmp
```
... where in.gif is a paletted image, out.bmp will contain the same palette. Is that what you want?

**Comment by Solstice245** [comment]
> IM doesn't contain CLI options for general manipulation of palettes. However, if we have a simple command...
> 
> ```
> magick in.gif out.bmp
> ```
> 
> ... where in.gif is a paletted image, out.bmp will contain the same palette. Is that what you want?

This produces BMPs whose dimensions match the respective GIF image bounds, but I need images whose dimensions always match the GIF's logical bounds.

**Comment by Solstice245** [comment]
> How about using -remap. Get the color from the color map of the GIF and create a colormap image of them, one pixel per color appended together either horizontally or vertically. Then use -remap with the colormap image on your bmp or in your original command line after the -colors 256.
> 
> Note, I am not sure why you need -colors 256 as GIF is limited to 256 colors. Also -colors 256 does not guarantee you will get 256 colors if there are fewer colors in the image.

I did try using `-remap`, however, this does not actually set the palette of the image index, it only changes color values. Maintaining the exact same palette is a critical part of a workflow in creating sprite sheets for an old game I'm modding, so anything less will **not** do. Also, the `-colors` argument was for the BMP which was not necessarily being saved with single byte color depth.

**Comment by fmw42** [comment]
Use -type palette rather than -colors ... (also -colors without +dither or -dither none before it will dither your output)

**Comment by bztsrc** [comment]
@Solstice245 Since this still hasn't been answered yet, may I suggest an alternate solution?

> Maintaining the exact same palette is a critical part of a workflow

I had exactly the same requirement, and I quickly put together a solution: [pngpal](https://gitlab.com/bztsrc/pngpal). It is nowhere near as featureful as ImageMagick, and it can only save PNG outputs, but it gets the job done. You can then use ImageMagick to convert the result into whatever format you want.

> creating sprite sheets for an old game I'm modding

And to solve exactly that, I've created yet another tool [sprsheet](https://gitlab.com/bztsrc/sprsheet). This also supports fixed palette (`-f` flag), however this only accepts PNG input and only works on sRGB colorspace (the other tool can read various palette formats, like GIF palette, GIMP's, MSPaint's, Adobe's etc. and it supports not just sRGB, but HSV, CYMK and CIELAB colorspaces etc.)

Maybe these tools will be useful to you too, if you can't get ImageMagick working.

Cheers,
bzt

**Reply by Solstice245** [reply]
I did end up getting what I needed done based on PNG input, and using Pillow (A python image manipulation library) for color palette conversion.
